Core/GameEngine/Source/Common/System/MiniDumper.cpp

```diff
@@ -0,0 +1,698 @@
+			}
+		}
+		break;
+	case IncludeThreadCallback:
+		// We want all threads except the dumping thread
+		if (input.IncludeThread.ThreadId == m_dumpThreadId)
+		{
+			output.ThreadWriteFlags &= (~ThreadWriteThread);
+		}
+		break;
+	case MemoryCallback:
+#ifndef DISABLE_GAMEMEMORY
+		do
+		{
+			// DumpMemoryObjects will set outputMemorySize to 0 once it's completed, signalling the end of memory callbacks
+			DumpMemoryObjects(output.MemoryBase, output.MemorySize);
+		} while ((output.MemoryBase == 0 || output.MemorySize == 0) && m_dumpObjectsState != COMPLETED);
+#else
+		output.MemoryBase = 0;
+		output.MemorySize = 0;
+#endif
+		break;
+	case ReadMemoryFailureCallback:
+		DEBUG_LOG(("MiniDumper::CallbackInternal: ReadMemoryFailure with MemoryBase=%llu, MemorySize=%lu: error=%u",
+			input.ReadMemoryFailure.Offset, input.ReadMemoryFailure.Bytes, input.ReadMemoryFailure.FailureStatus));
+		break;
+	case CancelCallback:
+		output.Cancel = FALSE;
+		output.CheckCancel = FALSE;
+		break;
+	}
+
+	return success;
+}
+
+#ifndef DISABLE_GAMEMEMORY
+void MiniDumper::DumpMemoryObjects(ULONG64& memoryBase, ULONG& memorySize)
+{
+	// m_dumpObjectsState is used to keep track of the current "phase" of the memory dumping process
+	// m_dumpObjectsSubState is used to keep track of the progress within each phase, and is reset when advancing on to the next phase
+	switch (m_dumpObjectsState)
+	{
+	case MEMORY_POOLS:
+	{
+		// Dump all the MemoryPool instances in TheMemoryPoolFactory
+		// This only dumps the metadata, not the actual MemoryPool contents (done in the next phase).
+		if (TheMemoryPoolFactory == NULL)
+		{
+			m_dumpObjectsState = MEMORY_POOL_ALLOCATIONS;
+			break;
+		}
+
+		Int poolCount = TheMemoryPoolFactory->getMemoryPoolCount();
+		//m_dumpObjectsSubState contains the index in TheMemoryPoolFactory of the MemoryPool that is being processed
+		if (m_dumpObjectsSubState < poolCount)
+		{
+			MemoryPool* pool = TheMemoryPoolFactory->getMemoryPoolN(m_dumpObjectsSubState);
+			if (pool != NULL)
+			{
+				memoryBase = reinterpret_cast<ULONG64>(pool);
+				memorySize = sizeof(MemoryPool);
+				++m_dumpObjectsSubState;
+			}
+			else
+			{
+				m_dumpObjectsSubState = poolCount;
+			}
+		}
+
+		if (m_dumpObjectsSubState == poolCount)
+		{
+			m_dumpObjectsSubState = 0;
+			m_dumpObjectsState = MEMORY_POOL_ALLOCATIONS;
+		}
+		break;
+	}
+	case MEMORY_POOL_ALLOCATIONS:
+	{
+		// Iterate through all the allocations of memory pools and containing blobs that has been done via the memory pool factory
+		// and include all of the storage space allocated for objects
+		if (TheMemoryPoolFactory == NULL)
+		{
+			m_dumpObjectsState = DMA_ALLOCATIONS;
+			break;
+		}
+
+		//m_dumpObjectsSubState is used to track if the iterator needs to be initialized, otherwise just a counter of the number of items dumped
+		if (m_dumpObjectsSubState == 0)
+		{
+			m_rangeIter = TheMemoryPoolFactory->cbegin();
+			++m_dumpObjectsSubState;
```

Comment:
Perhaps can get rid of `m_dumpObjectsSubState` by simply initializing `m_rangeIter = TheMemoryPoolFactory->cbegin()` before this callback?